### PR TITLE
Smart exec

### PR DESCRIPTION
**Pull request by parmentelat**
here's a first rough implementation for #292 

I'm calling it smart exec; it's bound to shift-enter, and allows to execute the current cell and then proceed to the next cell when the context permits

as of this code there is no config flag yet, at this point it is more a base for discussion

### CODE REVIEW

**Comment by parmentelat** [comment]
I had recorded a session as a gif, but it resulted in a file larger than 10mb, so..

the points I wanted to make in this recorded session were as follows

* all this makes sense only with auto_selection turned on, and that's not merged yet
which is also why I have not implemented a config flag yet

* when doing shift-enter on a cell X that is the last one in its fragment, right now it is stuck (i.e. shift enter does not move to the next cell); however is the case where the following fragment has been shown and we have moved the selection back on cell X, one could argue that shift-enter should then indeed move to the next cell. 

  this IMO would be the ideal behaviour but it is
   * of relative minor importance, and 
   * significantly harder to implement.

  And so I ruled it out for now.

**Comment by parmentelat** [comment]
I found a way to get this to work the way I described as ideal above, so I changed my mind and went for it

**Comment by parmentelat** [comment]
here's what it looks like - this requires auto-select and auto-select-fragment

rise-smart-exec2

**Comment by damianavila** [comment]
This is really exciting, thanks for the gif.

>all this makes sense only with auto_selection turned on, and that's not merged yet
which is also why I have not implemented a config flag yet

No problem... I need to test, review and merge the other PR. Hopefully I can do that in the next few days.

>I found a way to get this to work the way I described as ideal above, so I changed my mind and went for it

+1

> this requires auto-select and auto-select-fragment

I have not tested/reviewed yet, but how is the general behavior without these requirements? I am willing to make those a default thing if we can provide a unified experience. I would not like to have the smart-exec *only* when some specific options are configured.

**Comment by parmentelat** [comment]
To clarify:
The 2 features are vastly independent in terms of their implementation
 I was just saying that the gif had been recorded with auto-select turned on as well

**Comment by damianavila** [comment]
> I was just saying that the gif had been recorded with auto-select turned on as well

Ah... OK, thanks for the clarification.

File: rise/static/main.js [hunk 1]
```diff
@@ -77,8 +77,30 @@ Object.getPrototypeOf(Jupyter.notebook).get_cell_elements = function () {
...
+function get_slide_type(cell) {
+  var slide_type = (cell.metadata.slideshow || {}).slide_type;
+  return ( (slide_type === undefined) || (slide_type == '-')) ? '' : slide_type;
+}
+
+function is_slide(cell)    {return get_slide_type(cell) == 'slide';}
+function is_subslide(cell) {return get_slide_type(cell) == 'subslide';}
+function is_fragment(cell) {return get_slide_type(cell) == 'fragment';}
+function is_regular(cell)  {return get_slide_type(cell) == '';}
+
```
Comment (damianavila) [inline rise/static/main.js]: We should be reusing these in other places, but that's for the future...

File: rise/static/main.js [hunk 2]
```diff
@@ -111,12 +133,11 @@ function markupSlides(container) {
...
-    var i, cell, slide_type;
 
-    for (i=0; i < cells.length; i++) {
-        cell = cells[i];
-        slide_type = (cell.metadata.slideshow || {}).slide_type;
-        //~ console.log('cell ' + i + ' is: '+ slide_type);
+    for (var i=0; i < cells.length; i++) {
+        var cell = cells[i];
+        var slide_type = get_slide_type(cell);
+        //~ console.log(`cell ${i} is: ${slide_type}`);
```
Comment (damianavila) [inline rise/static/main.js]: This comment can be removed.

File: rise/static/main.js [hunk 3]
```diff
@@ -129,15 +150,18 @@ function markupSlides(container) {
                 // Start new subslide
                 current_fragment = subslide_section = new_subslide();
             } else if (slide_type === 'fragment') {
-                current_fragment = $('<div>').addClass('fragment')
+              // record the <div class='fragment'> element corresponding
+              // to each fragment cell in the 'fragment_div' attribute
```
Comment (damianavila) [inline rise/static/main.js]: Why you need to do this? I guess the answer is below.

File: rise/static/main.js [hunk 4]
```diff
@@ -157,6 +181,41 @@ function markupSlides(container) {
         if (slide_type === 'skip') {
             cell.element.addClass('reveal-skip');
         }
+
+      //console.log(`marked cell ${i} as ${cell.smart_exec}`)
```
Comment (damianavila) [inline rise/static/main.js]: Another debug thing that can be removed.

File: rise/static/main.js [hunk 5]
```diff
@@ -157,6 +181,41 @@ function markupSlides(container) {
...
+    }
+
+  // set on all cells a smart_exec tag that says how smart exec
+  // should behave on that cell
+  // the fragment cell also get a smart_exec_next_fragment
+  // attribute that points at the <div class='fragment'>
```
Comment (damianavila) [inline rise/static/main.js]: >Why you need to do this? I guess the answer is below.

And here is the answer :wink:

File: rise/static/main.js [hunk 6]
```diff
@@ -554,14 +640,6 @@ function reveal_cell_index(notebook) {
   // just scan all cells until we find one at that address
   // except that we need to start at -1 0r 0 depending on
   // whether the first slide has a slide tag or not
-  var is_slide = function(cell) {
```
Comment (damianavila) [inline rise/static/main.js]: Nice to see we are removing duplication.

**Comment by parmentelat** [comment]
I have taken care of the extra debug statements

As far as the conflict: this page here https://github.com/damianavila/RISE/pull/295/conflicts
shows me one conflict on line 697; but I don't understand how this is a conflict; the 3 old function definitions must be removed from there indeed..

I don't see how I can help out from my end, let me know if you need more info to merge this

**Comment by damianavila** [comment]
Try merging master into your branch... that should fix the issue, I think.

**Comment by parmentelat** [comment]
I need to go for today; let me know if you need anything else on my end for merging this

**Comment by damianavila** [comment]
> I need to go for today; let me know if you need anything else on my end for merging this

I think we are OK, thanks for your feedback.

**Comment by damianavila** [comment]
Btw, #308 add some comments about the functionality in the usage docs.

**Pull request merged**